apps/system/js/nfc_handover_manager.js

```diff
@@ -14,689 +13,716 @@
+          self.actionQueue = [];
+        };
+        req.onerror = function bt_getAdapterError() {
+          self._logVisibly('event listner: Failed to get bluetooth adater');
+        };
+      });
 
-    var cps = this.bluetooth.enabled ? NDEF.CPS_ACTIVE : NDEF.CPS_ACTIVATING;
-    var mac = this.defaultAdapter.address;
-    var hs = NDEFUtils.encodeHandoverSelect(mac, cps);
-    var promise = nfcPeer.sendNDEF(hs);
-    promise.then(() => {
-      this._debug('sendNDEF(hs) succeeded');
-      this.incomingFileTransferInProgress = true;
-    }).catch(e => {
-      this._logVisibly('sendNDEF(hs) failed : ' + e);
-      this._clearTimeout();
-      this._restoreBluetoothStatus();
-    });
-
-    this._clearTimeout();
-    this.responseTimeoutFunction =
-      setTimeout(this._cancelIncomingFileTransfer.bind(this),
-                 this.responseTimeoutMillis);
-  },
-
-  /**
-   * Initiate a file transfer by sending a Handover Request to the
-   * remote device.
-   * @param {MozNFCPeer} An instance of MozNFCPeer object.
-   * @param {Blob} blob File to be sent.
-   * @param {String} requestId Request ID.
-   * @memberof NfcHandoverManager.prototype
-   */
-  _initiateFileTransfer:
-    function _initiateFileTransfer(nfcPeer, blob, requestId) {
-      this._debug('initiateFileTransfer');
-      /*
-       * Initiate a file transfer by sending a Handover Request to the
-       * remote device.
-       */
+      window.addEventListener('bluetooth-disabled', function() {
+        self._debug('bluetooth-disabled');
+        self._clearBluetoothStatus();
+      });
+
+      window.navigator.mozSetMessageHandler('nfc-manager-send-file',
+        function(msg) {
+          self._debug('In New event nfc-manager-send-file' +
+            JSON.stringify(msg));
+          self.handleFileTransfer(msg.peer, msg.blob, msg.requestId);
+      });
+
+      SettingsListener.observe('nfc.debugging.enabled', false,
+                               (enabled) => { this.DEBUG = enabled; });
+
+      window.addEventListener('nfc-transfer-started',
+        this._transferStarted.bind(this)
+      );
+      window.addEventListener('nfc-transfer-completed',
+        this.transferComplete.bind(this)
+      );
+    },
+
+    /**
+     * Save the on/off status of Bluetooth.
+     * @memberof NfcHandoverManager.prototype
+     */
+    _saveBluetoothStatus: function _saveBluetoothStatus() {
+      if (!this.bluetoothStatusSaved) {
+        this.bluetoothStatusSaved = true;
+        this.bluetoothAutoEnabled = !this.bluetooth.enabled;
+      }
+    },
+
+    /**
+     * Restore the Bluetooth status.
+     * @memberof NfcHandoverManager.prototype
+     */
+    _restoreBluetoothStatus: function _restoreBluetoothStatus() {
+      if (!this.isHandoverInProgress() &&
+          BluetoothTransfer.isSendFileQueueEmpty) {
+        if (this.bluetoothAutoEnabled) {
+          this._debug('Disabling Bluetooth');
+          this.settings.createLock().set({'bluetooth.enabled': false});
```

Comment:
// XXX: Let BluetoothCore do this for you
this.writeSetting({'bluetooth.enabled': false});